lib/charms/bind/v0/dns_record.py

```diff
@@ -0,0 +1,570 @@
+
+    def get_remote_relation_data(self) -> Optional[DNSRecordProviderData]:
+        """Retrieve the remote relation data.
+
+        Returns:
+            DNSRecordProviderData: the relation data.
+        """
+        relation = self.model.get_relation(self.relation_name)
+        return self._get_remote_relation_data(relation) if relation else None
+
+    def _get_remote_relation_data(self, relation: ops.Relation) -> DNSRecordProviderData:
+        """Retrieve the remote relation data.
+
+        Args:
+            relation: the relation to retrieve the data from.
+
+        Returns:
+            DNSRecordProviderData: the relation data.
+        """
+        assert relation.app
+        relation_data = relation.data[relation.app]
+        return DNSRecordProviderData.from_relation_data(relation_data)
+
+    def _is_remote_relation_data_valid(self, relation: ops.Relation) -> bool:
+        """Validate the relation data.
+
+        Args:
+            relation: the relation to validate.
+
+        Returns:
+            true: if the relation data is valid.
+        """
+        try:
+            _ = self._get_remote_relation_data(relation)
+            return True
+        except ValidationError as ex:
+            error_fields = set(
+                itertools.chain.from_iterable(error["loc"] for error in ex.errors())
+            )
+            error_field_str = " ".join(f"{f}" for f in error_fields)
+            logger.warning("Error validation the relation data %s", error_field_str)
+            return False
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Event emitted when the relation has changed.
+
+        Args:
+            event: event triggering this handler.
+        """
+        assert event.relation.app
+        relation_data = event.relation.data[event.relation.app]
+        if relation_data and self._is_remote_relation_data_valid(event.relation):
+            self.on.dns_record_request_processed.emit(
+                event.relation, app=event.app, unit=event.unit
+            )
+
+    def update_relation_data(
+        self, relation: ops.Relation, dns_record_requirer_data: DNSRecordRequirerData
+    ) -> None:
+        """Update the relation data.
+
+        Args:
+            relation: the relation for which to update the data.
+            dns_record_requirer_data: a DNSRecordRequirerData instance wrapping the data to be
+                updated.
+        """
+        relation_data = dns_record_requirer_data.to_relation_data()
+        relation.data[self.charm.model.app].update(relation_data)
+
+
+class DNSRecordProvidesEvents(ops.CharmEvents):
+    """DNS record provider events.
+
+    This class defines the events that a DNS record provider can emit.
+
+    Attributes:
+        dns_record_request_received: the DNSRecordRequestReceived.
+    """
+
+    dns_record_request_received = ops.EventSource(DNSRecordRequestReceived)
+
+
+class DNSRecordProvides(ops.Object):
```

Comment:
Shouldn't we provide as part of the public API here helpers to set a status for a dns_domains or a dns_entries based on its uuid ?
Something like:
`set_domain_status(self, uuid: str, status: Status)`
And
`set_record_status(self, uuid: str, status: Status)`

Reply:
Yes, will do in a follow up PR